File: B_resources/b_Python_sklearn/lesion_encoder.py
import pandas as pd


def lesion_encode(df, ohe_cols, num_cols):
    df['numLesions'] = (df[['lesion_1', 'lesion_2', 'lesion_3']] != "0").sum(axis=1)
    num_cols = num_cols + ['numLesions']

    df['lesionSite'] = -1
    df['lesionType'] = -1
    df['lesionSubType'] = -1
    df['lesionCode'] = -1

    for i, row in df.iterrows():
        for lesion_col in ['lesion_1', 'lesion_2', 'lesion_3']:
            lesion_val = row[lesion_col]
            if lesion_val != "" and lesion_val != 'nan':
                if len(lesion_val) == 5:
                    if lesion_val[:2] == "11":
                        df.at[i, 'lesionSite'] = 11
                        df.at[i, 'lesionType'] = int(lesion_val[2])
                        df.at[i, 'lesionSubType'] = int(lesion_val[3])
                        df.at[i, 'lesionCode'] = int(lesion_val[4])
                    else:
                        if df.at[i, 'lesionSite'] < int(lesion_val[0]):
                            df.at[i, 'lesionSite'] = int(lesion_val[0])
                            df.at[i, 'lesionType'] = int(lesion_val[1])
                            df.at[i, 'lesionSubType'] = int(lesion_val[2])
                            df.at[i, 'lesionCode'] = int(lesion_val[3:5])
                else:
                    if df.at[i, 'lesionSite'] < int(lesion_val[0]):
                        lesion_val = lesion_val.ljust(4, '0')
                        df.at[i, 'lesionSite'] = int(lesion_val[0])
                        df.at[i, 'lesionType'] = int(lesion_val[1])
                        df.at[i, 'lesionSubType'] = int(lesion_val[2])
                        df.at[i, 'lesionCode'] = int(lesion_val[3])

    df.drop(['lesion_1', 'lesion_2', 'lesion_3'], axis=1, inplace=True)
    ohe_cols = [item for item in ohe_cols if item not in ['lesion_1', 'lesion_2', 'lesion_3']]

    lesionSiteDict = {"11": "all intestinal sites",
                      "1": "gastric",
                      "2": "sm intestine",
                      "3": "lg colon",
                      "4": "lg colon and cecum",
                      "5": "cecum",
                      "6": "transverse colon",
                      "7": "retum/descending colon",
                      "8": "uterus",
                      "9": "bladder",
                      "0": "none",
                      "-1": "n/a"}

    lesionTypeDict = {"1": "simple",
                      "2": "strangulation",
                      "3": "inflammation",
                      "4": "other",
                      "5": "other",
                      "6": "other",
                      "7": "other",
                      "8": "other",
                      "9": "other",
                      "0": "other",
                      "-1": "n/a"}

    lesionSubTypeDict = {"1": "mechanical",
                         "2": "paralytic",
                         "3": "n/a",
                         "4": "n/a",
                         "5": "n/a",
                         "6": "n/a",
                         "7": "n/a",
                         "8": "n/a",
                         "9": "n/a",
                         "0": "n/a",
                         "-1": "n/a"}

    lesionCodeDict = {"10": "displacement",
                      "1": "obturation",
                      "2": "intrinsic",
                      "3": "extrinsic",
                      "4": "adynamic",
                      "5": "volvulus/torsion",
                      "6": "intussuption",
                      "7": "thromboembolic",
                      "8": "hernia",
                      "9": "lipoma/slenic incarceration",
                      "0": "n/a",
                      "-1": "n/a"}

    df['lesionSite'] = df['lesionSite'].astype(str)
    df['lesionType'] = df['lesionType'].astype(str)
    df['lesionSubType'] = df['lesionSubType'].astype(str)
    df['lesionCode'] = df['lesionCode'].astype(str)

    df['lesionSite'] = df['lesionSite'].replace(lesionSiteDict)
    df['lesionType'] = df['lesionType'].replace(lesionTypeDict)
    df['lesionSubType'] = df['lesionSubType'].replace(lesionSubTypeDict)
    df['lesionCode'] = df['lesionCode'].replace(lesionCodeDict)

    return df, ohe_cols, num_cols

File: 02.resources/01.R/lesion_encoder.R
select <- dplyr::select  # select is masked by some libraries

lesionPrep <- function(data, non_ohe_cols) {
  
  data$lesion_1 <- data$lesion_1 %>% as.character()
  data$lesion_2 <- data$lesion_2 %>% as.character()
  data$lesion_3 <- data$lesion_3 %>% as.character()
  
  data$numLesions <- rowSums(data[c("lesion_1", "lesion_2", "lesion_3")] != "0")
  
  data$lesionSite <-  -1
  data$lesionType <- -1
  data$lesionSubType <- -1
  data$lesionCode <- -1
  
  non_ohe_cols <-  c(non_ohe_cols, 'numLesions')
  
  for (i in 1:nrow(data)){
    for (j in c("lesion_1", "lesion_2", "lesion_3")){
      lesion_val = data[i, j]
      if (!is.na(lesion_val) && !is.na(lesion_val) ){
        if (nchar(lesion_val) == 5) {
          if (substr(lesion_val, 1, 2) == "11"){
            data$lesionSite[i] <- 11
            data$lesionType[i] <- substr(lesion_val, 3, 3) %>% as.numeric()
            data$lesionSubType[i] <- substr(lesion_val, 4, 4) %>% as.numeric()
            data$lesionCode[i] <- substr(lesion_val, 5, 5) %>% as.numeric()
          }
          else {
            if (data$lesionSite[i] < substr(lesion_val, 1, 1) %>% as.numeric()){
              data$lesionSite[i] <- substr(lesion_val, 1, 1) %>% as.numeric()
              data$lesionType[i] <- substr(lesion_val, 2, 2) %>% as.numeric()
              data$lesionSubType[i] <- substr(lesion_val, 3, 3) %>% as.numeric()
              data$lesionCode[i] <- substr(lesion_val, 4, 5) %>% as.numeric()
            }
          }
        }
        else {
          if (data$lesionSite[i] < substr(lesion_val, 1, 1) %>% as.numeric()) {
            lesion_val <- str_pad(lesion_val, width = 4, side = "right", pad = "0")
            data$lesionSite[i] <- substr(lesion_val, 1, 1) %>% as.numeric()
            data$lesionType[i] <- substr(lesion_val, 2, 2) %>% as.numeric()
            data$lesionSubType[i] <- substr(lesion_val, 3, 3) %>% as.numeric()
            data$lesionCode[i] <- substr(lesion_val, 4, 4) %>% as.numeric()
          }
        }
      }
    }
  }
  
  data <- data %>% select(-c("lesion_1", "lesion_2", "lesion_3"))
  non_ohe_cols <- non_ohe_cols[!non_ohe_cols %in% c("lesion_1", "lesion_2", "lesion_3")]
  
  lesionSiteDict <- c("11" = "all intestinal sites",
                      "1" = "gastric",
                      "2" = "sm intestine",
                      "3" = "lg colon",
                      "4" = "lg colon and cecum",
                      "5" = "cecum",
                      "6" = "transverse colon",
                      "7" = "retum/descending colon",
                      "8" = "uterus",
                      "9" = "bladder",
                      "0" = "none",
                      "-1" = "none")
  
  lesionTypeDict <- c("1" = "simple",
                      "2" = "strangulation",
                      "3" = "inflammation",
                      "4" = "other",
                      "5" = "other",
                      "6" = "other",
                      "7" = "other",
                      "8" = "other",
                      "9" = "other",
                      "0" = "other",
                      "-1" = "other") 
  
  lesionSubTypeDict <- c("1" = "mechanical",
                         "2" = "paralytic",
                         "3" = "n/a",
                         "4" = "n/a",
                         "5" = "n/a",
                         "6" = "n/a",
                         "7" = "n/a",
                         "8" = "n/a",
                         "9" = "n/a",
                         "0" = "n/a",
                         "-1" = "n/a")   
  
  lesionCodeDict <- c("10" = "displacement",
                      "1" = "obturation",
                      "2" = "intrinsic",
                      "3" = "extrinsic",
                      "4" = "adynamic",
                      "5" = "volvulus/torsion",
                      "6" = "intussuption",
                      "7" = "thromboembolic",
                      "8" = "hernia",
                      "9" = "lipoma/slenic incarceration",
                      "0" = "n/a",
                      "-1" = "n/a")
  
  data <- data %>% mutate(lesionSite = 
                            str_replace_all(lesionSite, lesionSiteDict))
  
  data <- data %>% mutate(lesionType = 
                            str_replace_all(lesionType, lesionTypeDict))
  
  data <- data %>% mutate(lesionSubType = 
                            str_replace_all(lesionSubType, lesionSubTypeDict))
  
  data <- data %>% mutate(lesionCode = 
                            str_replace_all(lesionCode, lesionCodeDict))

  return(list(data, non_ohe_cols))
}
  
